hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java

```diff
@@ -0,0 +1,233 @@
+ */
+public class OMOpenKeysDeleteRequest extends OMKeyRequest {
+
+  private static final Logger LOG =
+          LoggerFactory.getLogger(OMOpenKeysDeleteRequest.class);
+
+  public OMOpenKeysDeleteRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long trxnLogIndex, OzoneManagerDoubleBufferHelper omDoubleBufferHelper) {
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumOpenKeyDeleteRequests();
+
+    OzoneManagerProtocolProtos.DeleteOpenKeysRequest deleteOpenKeysRequest =
+            getOmRequest().getDeleteOpenKeysRequest();
+
+    List<OpenKeyBucket> submittedOpenKeyBucket =
+            deleteOpenKeysRequest.getOpenKeysPerBucketList();
+
+    long numSubmittedOpenKeys = submittedOpenKeyBucket.stream()
+        .mapToLong(OpenKeyBucket::getKeysCount).sum();
+
+    LOG.debug("{} open keys submitted for deletion.", numSubmittedOpenKeys);
+    omMetrics.incNumOpenKeysSubmittedForDeletion(numSubmittedOpenKeys);
+
+    OzoneManagerProtocolProtos.OMResponse.Builder omResponse =
+            OmResponseUtil.getOMResponseBuilder(getOmRequest());
+
+    IOException exception = null;
+    OMClientResponse omClientResponse = null;
+    Result result = null;
+    Map<String, OmKeyInfo> deletedOpenKeys = new HashMap<>();
+
+    try {
+      // Open keys are grouped by bucket, but there may be multiple buckets
+      // per volume. This maps volume name to volume args to track
+      // all volume updates for this request.
+      Map<String, OmVolumeArgs> modifiedVolumes = new HashMap<>();
+      OMMetadataManager metadataManager = ozoneManager.getMetadataManager();
+
+      for (OpenKeyBucket openKeyBucket: submittedOpenKeyBucket) {
+        // For each bucket where keys will be deleted from,
+        // get its bucket lock and update the cache accordingly.
+        Map<String, OmKeyInfo> deleted = updateOpenKeyTableCache(ozoneManager,
+            trxnLogIndex, openKeyBucket);
+
+        deletedOpenKeys.putAll(deleted);
+
+        // If open keys were deleted from this bucket and its volume still
+        // exists, update the volume's byte usage in the cache.
+        if (!deleted.isEmpty()) {
+          String volumeName = openKeyBucket.getVolumeName();
+          // Returns volume args from the cache if the volume is present,
+          // null otherwise.
+          OmVolumeArgs volumeArgs = getVolumeInfo(metadataManager, volumeName);
+
+          // If this volume still exists, decrement bytes used based on open
+          // keys deleted.
+          // The volume args object being updated is a reference from the
+          // cache, so this serves as a cache update.
+          if (volumeArgs != null) {
+            // If we already encountered the volume, it was a reference to
+            // the same object from the cache, so this will update it.
+            modifiedVolumes.put(volumeName, volumeArgs);
```

Comment:
Just to clarify, is this the execution you are talking about?

1. Request1 deletes key1 from volume1 in cache.
2. Request2 deletes key2 from volume1 in cache.
3. Request1 sets cached VolumeArgs object volArgs.bytesUsed -= key1.bytesUsed.
    - *divergence 1*: The cache shows key1 and key2 as deleted, but cache byte usage only reflects key1's deletion.
4. Request2 sets cached VolumeArgs object volArgs.bytesUsed -= key2.bytesUsed.
    - At this point, byte usage in the cache is consistent with the keys it shows as deleted.
5. Response1 is processed, committing volArgs and the deletion of key1 to the DB.
    - *divergence 2*: the DB shows only key1 deleted, but volume byte usage has been set as if both key1 and key2 were deleted.
6. Response2 is processed, committing volArgs to the DB again, and committing the deletion of key2 to the DB.
    - Now the keys deleted and bytes used align in the DB.

IIRC the entire volume table is stored in memory and only persisted to the DB to save state. Reads only happen from the in memory cache for volume metadata. In this case, *divergence 2* will never be detected by callers since it only happens at the DB level. *divergence 1* my exist briefly and be detected by callers. Again, this is really an issue with all requests modified in HDDS-4053 and not just this PR. We should discuss to determine whether the slight inconsistency warrants a whole volume lock on all requests that modify byte usage.

Reply:
Not exactly, what I mean here is by the time we process request1, as we add the same object to double buffer, and if other thread processing request2 and updating it will update DB state also (Technically this should happen after adding response to double buffer)

Cache is for holding in flight updates if it is not committed to DB, I see no issues with that, this is by design.

Diveregence 1 should not exist, if volume locks are held.

Reply:
https://issues.apache.org/jira/browse/HDDS-2344 Jira but here just value updating (Might not be bringing ConcurrentModificationException, but it can provide some context)

Reply:
Thanks for the explanation @bharatviswa504. I now see that *divergence 2* in the above example poses an issue in the event of an OM crash happening between steps 5 and 6. This will cause the byte usage update to be applied twice in the DB after OM restart. Volume byte usage updates will be removed from the open key requests and responses. Since this is really a larger problem with all requests/responses operating in this way under HDDS-541, we can fix the issue when a solution is developed for all requests/responses.